### PR TITLE
`generate profile` command

### LINKED ISSUE #287: "nsc generate profile"
Given an URL of the form:

`nsc://operator/account/user` resolve a JSON that looks like:

```json
{
  "user_creds": "<filepath>",
  "operator" : {
     "service": "hostport"
   }
}
```

The URL can provide arguments:
- `keystore=<dir>` that specifies the location of the keystore
- `store=<dir>` that specifies a directory that contains the named operator
- `[user|account|operator]Key` - includes the public key for user, account, operator,  If no prefix (user/account/operator is provided, it targets the last object in the configuration path)
- `[user|account|operator]Seed=<optional public key>` - include the seed for user, account, operator, if an argument is provided, the seed for the specified public key is provided - this allows targeting a signing key. If no prefix (user/account/operator is provided, it targets the last object in the configuration path)
- `[user|account|operator]Name` - includes the friendly name for the for user, account, operator,  If no prefix (user/account/operator is provided, it targets the last object in the configuration path)

`nsc://operator/account/user?seed`:
```json
{
  "operator": { "service": "host:port" },
  "user": {
    "seed": "SUXXXXXXXXX"
  }
}
```

NSC will offer an interface where it can request it as API, but will also allow you to do: `nsc generate profile <URL>` and spew the JSON out which can be redirected using `--out-file`.

### PR DESCRIPTION
**Pull request by aricart**
Added an internal `generate profile` mechanism to generate a JSON with interesting bits for tooling
FIX #287

### CODE REVIEW

**Comment by aricart** [comment]
@ripienaar still in wip, needs more tests, is there anything else you would like in this?

File: cmd/resolve.go [hunk 1]
```diff
@@ -0,0 +1,596 @@
...
+
+type Details struct {
+	Service []string `json:"service,omitempty"`
+	Name    string   `json:"name,omitempty"`
+	Seed    string   `json:"seed,omitempty"`
+	Key     string   `json:"id,omitempty"`
+}
+
+type ResolveResults struct {
+	UserCreds string   `json:"userCreds,omitempty"`
```
Comment (ripienaar) [inline cmd/resolve.go:110]: `user_creds` surely? :)

Comment (aricart) [reply cmd/resolve.go:110]: lol - that snake_casing_bad....

Comment (ripienaar) [reply cmd/resolve.go]: Just cos JS people cant make up their mind doesnt make it bad :P

Comment (aricart) [reply cmd/resolve.go]: lolz

**Comment by aricart** [comment]
@ripienaar @matthiashanel renamed the command `profile` and hung it from `generate` - this seems more in line with it's intended use

**Comment by ripienaar** [comment]
LGTM :)

File: cmd/resolve.go [hunk 2]
```diff
@@ -0,0 +1,672 @@
...
+Other options (as query string arguments):
+keystore=<dir> that specifies the location of the keystore
+
+store=<dir> that specifies a directory that contains the named operator
+
+[user|account|operator]Key - includes the public key for user, account, 
+operator, If no prefix (user/account/operator is provided, it targets 
+the last object in the configuration path)
+
+[user|account|operator]Seed=<optional public key> - include the seed for
```
Comment (matthiashanel) [inline cmd/resolve.go]: list an example with the public key? I also can't find a unit test for this

File: cmd/resolve.go [hunk 3]
```diff
@@ -0,0 +1,672 @@
...
+
+func (u *NscURL) getAccount() (string, error) {
+	return url.QueryUnescape(u.account)
+}
+
+func (u *NscURL) getUser() (string, error) {
+	return url.QueryUnescape(u.user)
+}
+
+func (u *NscURL) query() (map[Arg]string, error) {
```
Comment (matthiashanel) [inline cmd/resolve.go]: This looks like functionality that can be found in "net/url"
https://golang.org/pkg/net/url/#ParseQuery

File: cmd/resolve.go [hunk 4]
```diff
@@ -0,0 +1,672 @@
...
+				return nil, err
+			}
+			v = s
+		}
+		m[Arg(k)] = v
+	}
+	return m, nil
+}
+
+func ParseNscURL(u string) (*NscURL, error) {
```
Comment (matthiashanel) [inline cmd/resolve.go]: I assume this id done manually as we don't use a host and so url parsing may be tricky?

File: cmd/resolve.go [hunk 5]
```diff
@@ -0,0 +1,672 @@
...
+	names = append(names, cd.Subject)
+
+	conf := GetConfig()
+	payload := c.Payload()
+	_, ok := payload.(jwt.Operator)
+	if ok && conf.Operator != "" {
+		names = append(names, conf.Operator)
+	}
+	return names
+}
```
Comment (matthiashanel) [inline cmd/resolve.go]: map[string]struct{}?

File: cmd/resolve.go (hunk 2 above)
Comment (aricart) [reply cmd/resolve.go]: done

File: cmd/resolve.go (hunk 3 above)
Comment (aricart) [reply cmd/resolve.go]: Parse query is broken for arguments that don't have a value:
func (v Values) Get(key string) string {
	if v == nil {
		return ""
	}
	vs := v[key]
	if len(vs) == 0 {
		return ""
	}
	return vs[0]
}

File: cmd/resolve.go (hunk 4 above)
Comment (aricart) [reply cmd/resolve.go]: correct - I am allowing spaces or URL encoded.

File: cmd/resolve.go (hunk 5 above)
Comment (aricart) [reply cmd/resolve.go]: done

File: cmd/generateprofile.go [hunk 6]
```diff
@@ -0,0 +1,680 @@
...
+type NscURL struct {
+	operator string
+	account  string
+	user     string
+	qs       string
+}
+
+type stringSet struct {
+	set map[string]string
+}
```
Comment (matthiashanel) [inline cmd/generateprofile.go:190]: type stringSet map[string]struct{} would have worked as well. Presence is the only thing checked.

**Review by matthiashanel — APPROVED** [review]
LGTM

assuming `resolve nsc://operator?userKey` would raise an error.

**Pull request merged**